README.md
# Insane Noises Music Generator

This is the final group project done for COSC-452 Seminar in Computer Science: Evolutionary Computation at Amherst College where we were allowed to do anything as long as it utilized evolutionary computation principles. The code was written in collaboration with Lemara Williams, Class of 2022, and April Dotton-Carter, Class of 2023 (this README for my personal GitHub is being written by myself, Karen Liu, Class of 2023). Further details about what code was written by myself is provided below in the Personal Contributions section.

## Contents

-   Overview: Aims, Activities, Products

-   Detailed Description of the System

-   Personal Contributions and Future Work

## Overview: Aims, Activities, Products

Our group set out to create a music evolution system that could seamlessly integrate with a visual simulation that animates the music it is given, the best final individual from the system. Our activities can be broken down into two major components --- working on the evolutionary system and working on the visual simulation. These two aspects worked in parallel. While the evolutionary half progressed steadily from the beginning based on Lemara's original code, the visual simulation half of things had two major stages --- coming up with ideas for the visual and then actually coding it. Our results are a functional system that produces music, produces a visual, and can essentially stand alone, though of course there is much more that can be done --- for instance, we should test more parameters that affect variation and add more cases to the error function in order to make "better" music; we could also consider a wholly different visual system that may achieve better intuitive connection to the music than our current one.

## Detailed Description of the System

The evolutionary system starts with 50 individuals of genome length 65 that will go through 100 generations of variation and selection. Each gene in the individuals is a random MIDI note between 36 and 84, with a random duration of ½, 1, 2, or 4 to represent eighth notes, quarter notes, half notes, and whole notes.

The variation method is a combination of mutation and crossover. The mutation is a looser version of UMAD in that it does retain uniformity over time, but rather than performing both an addition and a deletion sweep in the same generation, there is a 50% chance for each addition and deletion so that only one type of sweep is performed per generation. This allows for greater fluctuations in genome length, which we hoped would be helpful for keeping the sounds interesting. The crossover function is essentially the one in Propeller, but it swaps genes 50% of the time, which is a higher frequency than in Propeller.

The selection method is lexicase selection, and our error function tests for genome length on a hyperbolic shape so that genome lengths will be penalized dramatically if they are shorter than or closer to 30. This way, the error will never be zero, which is why our system always runs through the max number of generations.

```
# Lexicase selection
(if (or (<= (count genome) 30)
        (> (count genome) 100))
  10000000
  (/ 10000.0 (- (count genome) 30)))
```

The other characteristic our error function tests for is the note intervals, penalizing tritones, octave jumps, and intervals over a fifth a certain percentage of the time and leaving them unpenalized the rest of the time. The reason we decided to penalize these intervals is because we wanted to abide by classical music conventions as an initial basis for achieving "pleasing" sounds since there is no other clear set of rules we could implement in code without going way beyond the scope of a single month. In order to avoid a boring, strict adherence to these rules, we implemented them so that these rules are not penalized all the time in order to allow for some intrigue.

The visual system aims to create an animation of contracting and expanding circles that vaguely imitate the pulsing of a loud speaker. Each characteristic is keyed to an element of the visual, like so:

-   Radius MIDI number, or pitch

-   Color (Hue) MIDI number, or pitch

-   Color (Brightness) How many times that same note has been hit

-   Duration of contraction Duration of note

The circle will appear at the note's location and leave a colored "imprint" of itself at the beginning of every note and it will contract until the beginning of the next note. When the same note is hit again, the imprint will become brighter and brighter. The final product should look like almost like a record, with a clear, colored distribution of circle imprints that show what the overall "shape" of the song was like. These images can be seen in the presentation slides, which are attached as a separate PDF. Videos of some of the more interesting runs are also included.

## Personal Contributions and Future Work

My main roles included creating the visual half of the system, debugging and making modifications to the evolutionary half of the system, and merging the two systems together after we'd finished our respective halves; I also did all the testing and tweaking of the merged system, recorded the visual results of runs, and organized the results and future work sections of the presentation.

For the visual half of the system, I created the code from scratch based on my past exploration of the bouncing circles animation in Quil and wrote the structure of the entire file. It's structured so that every single characteristic, listed earlier, and the necessary information to code these visual qualities are keys in a map, and this map is the one that is passed from (setup [ ]) and between (update-state [state]) and (draw [state]). The only code that wasn't written by me were the keys and draw commands related to the brightness of the circles (which represent the amount of times that note has been hit).

For the evolutionary system, I debugged and cleaned up the code, which is the version in this repo. I also modified the error function to penalize genome length hyperbolically instead of with an absolute cutoff of which lengths are desirable and which are not. For variation, I created the mutation function based on old code from your class examples and added the crossover function to the system from Propeller.

Merging the system was difficult, as we ran into problems with getting Overtone to work after it had combined with the visual system. A teammate and I worked together on merging the system, and I eventually put Dynne in since I'd used it in past projects and Overtone did not seem to have any hope of working.

Testing was straightforward --- it was just manipulating many of the changeable parameters to approach a more "pleasing" set of final individuals. Parameters I tested included:

-   Initial genome size

-   Minimum desirable genome size in the error function

-   Ratio of crossover to mutation

-   Frequency of addition and deletion within mutation

-   Frequency of gene swapping in crossover

-   Frequency of allowing tritones, octaves, and jumps over a fifth to go unpenalized

Other things I want to test in the future, including the ones I've already listed:

-   Generation number

-   Initial population size

I'll include in the submission photos and videos that were in the presentation but couldn't be played during class due to technical difficulties. I won't reiterate future work because it was already talked about on the last slide of the presentation.

Other notes on future work and observations that weren't included in the presentation, however, include:

-   The best genome is always too short, and it seems like the system is prioritizing size over intervals.

-   Changing the color to key to the order in which it was drawn would preserve the single key to single value standard, and it may also show how the song plays over time; a time-tracked history of the notes played is not something currently seen in the final visual if only looking at the last image as opposed to the entire animation.

-   Long genomes look cooler but take forever to draw.

-   Batch testing on a cluster would be good if we had analysis scripts, but there's the question of how to get analysis scripts to capture the nuances of human evaluation and aesthetic judgment.

-   Unsure of how to control for genome length without being too restrictive or too permissive.

-   We should include in the error function a test on the variance of note lengths as well as note frequencies to modulate skips and steps, as well as rhythmic interest.

## Usage

Downloading the repo and running `core.clj` in the repo will produce a window that plays sound and shows the ring animation.

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-03-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-03-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/insane-noises/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/insane-noises/compare/0.1.0...0.1.1


src/insane_noises/circles.clj
(ns insane-noises.circles
  (:require [quil.core :as q :include-macros true]
            [quil.middleware :as m]
            [insane-noises.core :as core]
            ;[overtone.live :as o]
            ;[overtone.inst.piano]
            ))

(require '[dynne.sampled-sound :as d])

(defn midi-to-hz
  [m]
  (if (< m 21)
    0
    (* 440 (Math/pow 2
                     (-> m (- 69) (/ 12.0))))))

(defn scale [x min max a b]
  (+ a
     (-> (- b a)
         (* (- x min))
         (/ (- max min)))))

(defn scale-through [coll a b]
  (let [min (apply min coll)
        max (apply max coll)]
    (map #(scale % min max a b) coll)))



; count-occurrences function was copied from
; https://stackoverflow.com/questions/25614540/clojure-how-to-count-occurrences-in-a-list/25616080#25616080
(defn count-occurrences [obj vec]
  (->> vec
       (filter #{obj})
       count))

(defn setup []

  (q/color-mode :hsb)

  ;(let [results [60 60 67 67 69 69 67 65 65 64 64 62 62 60]
  ;      twinkle [60 60 67 67 69 69 67 65 65 64 64 62 62 60]
  ;      scaled-twinkle (scale-through twinkle 200.0 800.0)
  ;      durations [1.0 1.0 1.0 1.0 1.0 1.0 2.0 1.0 1.0 1.0 1.0 1.0 1.0 2.0]
  ;      colors (map #(Math/floor %)
  ;                  (scale-through twinkle 0 185))]

  (let [results (core/gp 50 100)
        genome (:genome results)
        twinkle (vec (map first genome))
        scaled-twinkle (scale-through twinkle 200.0 800.0)
        durations (vec (map second genome))
        colors (map #(Math/floor %)
                    (scale-through twinkle 0 185))]

    (clojure.pprint/pprint (assoc results :length (count (:genome results))))

    {:song                 twinkle
     :scaled-song          scaled-twinkle
     :all-colors           colors

     :note                 (first twinkle)
     :duration             (first durations)
     :radius               (first scaled-twinkle)
     :color                (first colors)
     :brightness           75
     :brightness-increment (Math/floor (/ 180 (reduce max (vals (frequencies twinkle)))))
     :frames               0

     :prev-notes           [(first scaled-twinkle)]
     :prev-colors          [(first colors)]
     :prev-brightnesses    [55]
     :prev-count           1

     :next-real-notes      (next twinkle)
     :next-notes           (next scaled-twinkle)
     :next-durs            (next durations)
     :next-colors          (next colors)

     :last                 false
     :player               (d/play (d/sinusoid 50.0 (midi-to-hz (first twinkle))))}))

(defn update-state [state]

  (if (< (Math/abs
           (- (* 2.5 (:duration state))
              (:frames state)))
         0.1)

    (if (nil? (first (:next-real-notes state)))

      (assoc state :last true)

      (let []
        (d/stop (:player state))
        {:song                 (:song state)
         :scaled-song          (:scaled-song state)
         :all-colors           (:all-colors state)

         :note                 (first (:next-real-notes state))
         :duration             (first (:next-durs state))
         :radius               (first (:next-notes state))
         :color                (first (:next-colors state))
         :frames               0

         :brightness-increment (:brightness-increment state)
         :brightness           (+ 55 (* (:brightness-increment state)
                                        (count-occurrences (first (:next-notes state))
                                                           (:prev-notes state))))
         :prev-brightnesses    (conj (:prev-brightnesses state)
                                     (+ 55 (* (:brightness-increment state)
                                              (count-occurrences (first (:next-notes state))
                                                                 (:prev-notes state)))))

         :prev-notes           (conj (:prev-notes state)
                                     (first (:next-notes state)))
         :prev-colors          (conj (:prev-colors state)
                                     (first (:next-colors state)))
         :prev-count           (+ 1 (:prev-count state))

         :next-real-notes      (next (:next-real-notes state))
         :next-notes           (next (:next-notes state))
         :next-durs            (next (:next-durs state))
         :next-colors          (next (:next-colors state))

         :last                 false
         :player               (d/play (d/sinusoid 50.0
                                                   (midi-to-hz (first (:next-real-notes state)))))}
        ))

    (assoc state :radius (- (:radius state) 1.0)
                 :frames (+ 0.1 (:frames state)))))

(defn draw-state
  [state]
  (q/background 0)
  (q/no-fill)

  (dotimes [n (:prev-count state)]
    (let [old-notes (:prev-notes state)
          old-colors (:prev-colors state)
          old-brightnesses (:prev-brightnesses state)
          r (nth old-notes n)
          color (nth old-colors n)
          brightness (nth old-brightnesses n)]
      (q/stroke color 255 brightness)
      (q/stroke-weight 2)
      (q/ellipse 750 450 r r)))

  (if (:last state)
    (let []
      (d/stop (:player state)))
    (let [r (:radius state)]
      (q/stroke 255)
      (q/stroke-weight 3)
      (q/ellipse 750 450 r r))))

; leave behind an imprint of the note, and make the color brighter
; every time it is hit again and have the color hue depend on the frequency

(q/defsketch mysketch
             :host "host"
             :size [1520 920]                               ;1520
             :setup setup
             :update update-state
             :draw draw-state
             :middleware [m/fun-mode])

src/insane_noises/row_row_row/row_row_row.clj
(ns insane-noises.row-row-row
  (:use [overtone.live]))

(defn string
  [freq duration]
  (with-overloaded-ugens
    (* (line:kr 1 1 duration FREE)
       (pluck (* (white-noise) (env-gen (perc 0.001 5) :action FREE))
              1 1 (/ 1 freq) (* duration 2) 0.25))))

(definst harpsichord [freq 440]
         (let [duration 1
               snd  (string freq duration)
               t1   (* 0.2 (string (* 2/1 freq) duration))
               t2   (* 0.15 (string (* 3/2 freq) duration))
               t3   (* 0.1 (string (* 4/3 freq) duration))
               t4   (* 0.1 (string (* 5/4 freq) duration))
               snd  (+ snd (mix [t1 t2 t3 t4]))]
           snd))

(def melody
  (let [pitches
        [67 67 67 69 71             ; Row, row, row your boat,
         71 69 71 72 74             ; Gently down the stream,
         79 79 79 74 74 74 71 71 71 ; Merrily, merrily, merrily, merrily,
         67 67 67 74 72 71 69 67]   ; Life is but a dream!
        durations
        [1 1 2/3 1/3 1
         2/3 1/3 2/3 1/3 2
         1/3 1/3 1/3 1/3 1/3 1/3 1/3 1/3 1/3 1/3 1/3 1/3
         2/3 1/3 2/3 1/3 2]
        times (reductions + 0 durations)]
    (map vector times pitches)))

(defn play [metro notes]
  (let [play-note (fn [[beat midi]] (at (metro beat) (-> midi midi->hz harpsichord)))]
    (dorun (map play-note notes))))

(defn after [beats metro] (comp metro #(+ % beats)))

(defn play-round [metro notes]
  (play metro notes)
  (play (after 4 metro) notes)
  (play (after 8 metro) notes)
  (play (after 16 metro) notes))

;(play (metronome 120) melody)
;(play-round (metronome 120) melody)

(defn -main [& args]
  (let [metro (metronome 120)]

    ;; play the melody
    (play metro melody)
    (play-round (after 16 metro) melody)

    ;; wait for the melody to finish
    (Thread/sleep (- (metro 48) (now)))))


src/insane_noises/core.clj
(ns insane-noises.core)

;;make a diverse "song" with different tones
(defn random [min max]
  (+ (rand-int (- max min)) min))

;(def foo [b] (synth (out 0 (pan2 (sin-osc b))))) ;just creates a synth
;
;(defsynth my-sin [freq 440 sustain 1]
;          (out 0 (pan2 (sin-osc freq))))
;; like fn and defn, always need a value assigned or uses default
;
;(defn play [note duration]
;  (let [time (now)]
;    (if (= note -1)
;      (at (time * duration 2000) (my-sin 0))
;      (at (time * duration 2000) (piano note)))))



;;;Our ingredients
;randomly generated ~30 individuals
;creates a gene with freq + duration (actual length of notes)
(defn ram [] (vector (random 36 84) (first (shuffle [1/2 1 2 4]))))

;;;;;;;Error Function Helpers
(defn absolute [n]
  (if (< n 0)
    (- n)
    n))

(defn pair-error [[[p1 d1] [p2 d2]]]
  (if (or (= p1 (- 12 p2))                                  ;;octave
          (= p1 (+ 12 p2))
          (= p1 p2))
    (if (>= (rand) 0.5)                                     ;;50%
      200
      0)
    0)
  (if (= p1 (+ 6 p2))                                       ;;tritone
    (if (<= (rand) 0.6)                                     ;;30%
      200
      0)
    0)
  (if (> (- p1 p2) 7)                                       ;; over a fifth
    (if (<= (rand) 0.9)                                     ;;10%
      200
      0)
    0))

;;;;Error Function

(defn error [genome]
  (+ (reduce + (map pair-error (partition 2 1 genome)))
     (if (or (<= (count genome) 30)
             (> (count genome) 100))
       10000000
       (/ 10000.0 (- (count genome) 30)))))

;; the error will tend to be large, bc as long as the count is in
;; between 30-100 the error won't reach 0
;;;;;;;


(defn new-individual []
  "Returns a new, random individual in the context of test-pairs."
  (let [genome (vec (repeatedly 65 ram))]
    {:genome genome
     :error  (error genome)}))

(defn best [individuals]
  "Returns the best of the given individuals."
  (reduce (fn [i1 i2]
            ;(print i1, i2)
            (if (< (:error i1) (:error i2))
              i1
              i2))
          individuals))

(defn lexicase-selection
  "Selects an individual from the population using lexicase selection."
  [pop]
  (loop [survivors (map rand-nth (vals (group-by :errors pop)))
         cases (shuffle (range (count (:errors (first pop)))))]
    (if (or (empty? cases)
            (empty? (rest survivors)))
      (rand-nth survivors)
      (let [min-err-for-case (apply min (map #(nth % (first cases))
                                             (map :errors survivors)))]
        (recur (filter #(= (nth (:errors %) (first cases)) min-err-for-case)
                       survivors)
               (rest cases))))))

(defn mutate [individual]
  "Returns a possibly-mutated copy of genome."
  (let [genome (:genome individual)
        with-additions (flatten (for [g genome]
                                  (if (< (rand) 0.3)
                                    (shuffle (list g (ram))) ;not using ingredients list
                                    g)))
        with-deletions (flatten (for [g genome]
                                  (if (< (rand) 0.3)
                                    ()
                                    g)))]
    (if (< (rand) 0.5)
      (partition 2 with-additions)
      (partition 2 with-deletions))))

(defn crossover                                             ;taken from Propeller
  "Crosses over two individuals using uniform crossover. Pads shorter one."
  [genome1 genome2]
  (let [plushy-a (:genome genome1)
        plushy-b (:genome genome2)
        shorter (min-key count plushy-a plushy-b)
        longer (if (= shorter plushy-a)
                 plushy-b
                 plushy-a)
        length-diff (- (count longer) (count shorter))
        shorter-padded (concat shorter (repeat length-diff :crossover-padding))]
    (remove #(= % :crossover-padding)
            (map #(if (< (rand) 0.5) %1 %2)
                 shorter-padded
                 longer))))

(defn make-child [population]
  "Returns a new, evaluated child, produced by mutating the result
  of crossing over parents that are selected from the given population."
  (if (< (rand) 1.0)
    (let [new-genome (mutate (lexicase-selection population))] ;breaks when making a child
      {:genome new-genome
       :error  (error new-genome)})
    (let [new-genome (crossover (lexicase-selection population)
                                (lexicase-selection population))]
      {:genome new-genome
       :error  (error new-genome)})))

(defn report [generation population]
  "Prints a report on the status of the population at the given generation."
  (let [current-best (best population)]
    (println {:generation   generation
              :best-error   (:error current-best)
              :diversity    (float (/ (count (distinct population))
                                      (count population)))
              :average-size (float (/ (->> population
                                           (map :genome)
                                           (map count)
                                           (reduce +))
                                      (count population)))
              :best-genome  (:genome current-best)})))

(defn gp [population-size generations]
  "Runs genetic programming to solve, or approximately solve, a floating-point
  symbolic regression problem in the context of the given population-size,
  number of generations to run, and test-pairs."
  (loop [population (repeatedly population-size
                                #(new-individual))
         generation 0]
    (report generation population)
    (if (or (< (:error (best population)) 0.1)
            (>= generation generations))
      (best population)
      (recur (conj (repeatedly population-size
                               #(make-child population)))
             (inc generation)))
    ))
